doc/source/usage.rst

```diff
@@ -4,17 +4,43 @@ Usage
 API overview
 ++++++++++++
 
+A Filehandle (:c:type:`DcmFilehandle`) enables access of a `DICOM file
+<http://dicom.nema.org/medical/dicom/current/output/chtml/part10/chapter_3.html#glossentry_DICOMFile>`_,
+which contains an encoded Data Set representing a SOP Instance.
+A Filehandle can be created via :c:func:`dcm_filehandle_create_from_file()`
+or :c:func:`dcm_filehandle_create_from_memory()` , and destroyed via
+:c:func:`dcm_filehandle_destroy()`.  You can make your own load functions
+to load from other IO sources, see :c:func:`dcm_filehandle_create()`.
+
+The content of a Part10 file can be read using various functions.
+
+The `File Meta Information
+<http://dicom.nema.org/medical/dicom/current/output/chtml/part10/chapter_3.html#glossentry_FileMetaInformation>`_
+can be accessed via :c:func:`dcm_filehandle_get_file_meta()`.
+
+The principal metadata of the Data Set can be accessed via
+:c:func:`dcm_filehandle_get_metadata()`. This function will stop read on tags
+which are likely to take a long time to process.
+
+You can read all metadata and control read stop using a sequence of calls to
+:c:func:`dcm_filehandle_read_metadata()`.
```

Comment:
It's a speed thing -- we need to stop parsing as soon as we see a tag that might potentially be dodgy, otherwise it's too late and we're committed to reading something horrible like `TAG_REFERENCED_IMAGE_NAVIGATION_SEQUENCE`.

Reply:
I have little experience with `TAG_REFERENCED_IMAGE_NAVIGATION_SEQUENCE`, but guess that it's similarly bad as `TAG_PER_FRAME_FUNCTIONAL_GROUP_SEQUENCE`.

While these sequence elements are annoying and take time to parse, their content is considered part of the metadata and are generally needed to understand the pixel data element. Therefore, I suggest reading them along with the rest of the metadata.

Reply:
You can use `/usr/bin/time -f %M:%e ...` to get peak ram use for a command, which can be very useful. With git master libdicom I see:

```
john@banana ~/pics/openslide/dicom/272320e7-5abb-ccac-8f8b-9c2008f53845_114626 $ /usr/bin/time -f %M:%e openslide-show-properties 1.3.6.1.4.1.36533.1214640142408762111756320324051133413.dcm > log
115148:0.06
```

So 115mb of memory and 60ms. If I enable the parsing of these two elements I see:

```
john@banana ~/pics/openslide/dicom/272320e7-5abb-ccac-8f8b-9c2008f53845_114626 $ /usr/bin/time -f %M:%e openslide-show-properties 1.3.6.1.4.1.36533.1214640142408762111756320324051133413.dcm > log
995708:8.91
```

Almost 1gb of memory use, and 9s. That's more than 100x slower, almost 10x more memory use, and no benefit for openslide.

Reply:
I think it's critical that `dcm_filehandle_get_metadata` reads all metadata by default. Therefore, I suggest changing the behavior of `dcm_filehandle_get_metadata`, such that the function reads all metadata, including `TAG_REFERENCED_IMAGE_NAVIGATION_SEQUENCE` and `TAG_PER_FRAME_FUNCTIONAL_GROUP_SEQUENCE`.

If an applications or an other library (e.g., OpenSlide) wants to read metadata differently for improved performance, it can use the underlying `dcm_filehandle_read_metadata` function directly rather than relying on `dcm_filehandle_get_metadata`.

Does this make sense?

Reply:
It actually does this now, but the other way around.

libdicom (and openslide) need to be able to open DICOMs quickly so they can figure out which files go together to make up a slide, and so that libdicom can handle things like tile lookup efficiently. Therefore, libdicom keeps an internal copy of a subset of the metadata for its own use, and makes that internal copy of the subset available to openslide via `dcm_filehandle_get_metadata()`.

You don't free the result of `dcm_filehandle_get_metadata()` any more (this is a change from the older version of the API), since libdicom is letting you access part of its internal state.

There's a new call, `dcm_filehandle_read_metadata()`, which really reads in metadata, and which the caller must free. That defaults to reading everything (except pixeldata), as you say.

Reply:
My main point is that `dcm_filehandle_get_metadata` should really provide all metadata.

I would expect the following behavior: When calling `dcm_filehandle_get_metadata`, I get all the metadata (i.e., all data elements that are not bulk data elements - note that SQ elements are per definition not bulk data elements). When called the first time, `dcm_filehandle_get_metadata` may call `dcm_filehandle_read_metadata` internally, but should do so without including `TAG_REFERENCED_IMAGE_NAVIGATION_SEQUENCE` and `TAG_PER_FRAME_FUNCTIONAL_GROUP_SEQUENCE` in `stop_tags` (see [here](https://github.com/jcupitt/libdicom/blob/3d44c8ad2c4486ff3a49e0747582752ccbae14e0/src/dicom-file.c#L766-L767)).

I consider this critical. The Google Healthcare API tried to be smart about that, too, and now fails to include the Per-Frame Functional Groups Sequence into metadata beyond a certain size, making it impossible to visualize larger images with TILED_SPARSE dimension organization.

We should make sure the default behavior is correct, even if it means that performance is slow.

Reply:
Couldn't OpenSlide use `dcm_filehandle_read_metadata` to selective read a subset of the metadata (without the Image Navigation Sequence or Per-Frame Functional Groups Sequence) that is needed to build the pyramid structure and then subsequently read the remaining metadata when necessary?

Reply:
For the `dcm-dump` tool, we could introduce a `--stop-tag` option that enables users to limit reading of metadata.

Reply:
Hmmm I'm not sure I understand. If you call `dcm_filehandle_read_metadata()` you get all the metadata by default. It's called `_read()` because it's a read function -- it fetches the next few tags (until it hits a stop tag) from the input file's current read position.

`dcm_filehandle_get_metadata()` is something different: it does not read in metadata, instead it's a simple (ie. *always* instant, no delay at all, ever) getter function that gives callers access to libdicom's internal copy of a subset of the file metadata.

Reply:
> For the dcm-dump tool, we could introduce a --stop-tag option that enables users to limit reading of metadata.

I agree, that would be a nice addition (`dcmdump` has something similar).

Reply:
I've done the rename.

Reply:
I really think these kind of performance optimizations belong into OpenSlide.

Providing the caller a (poorly defined) subset of the metadata is in my opinion against our core principle of providing a library that is free of surprises. The library includes functions that allow an application to load metadata in a more sophisticated, performance-optimized manner if needed. We can give examples in the documentation on how this can be achieved.

Reply:
` dcm_filehandle_get_metadata_subset()` is an optional call which can help performance, so to my mind it seems OK. 

The main metadata API call (`dcm_filehandle_read_metadata()`) is very surprise-free, don't you think? One call, reads in all the metadata.